File: survival-graphics/src/main/java/net/survival/graphics/CompositeDisplay.java
package net.survival.graphics;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.stream.Collectors;

import org.joml.Matrix4f;

import net.survival.block.Column;
import net.survival.graphics.opengl.GLMatrixStack;
import net.survival.graphics.opengl.GLRenderContext;
import net.survival.graphics.opengl.GLState;
import net.survival.graphics.particle.ClientParticleSpace;
import net.survival.render.ModelType;

public class CompositeDisplay implements GraphicsResource {
    private static final String ASSET_ROOT_PATH = "./assets/";
    private static final String TEXTURE_PATH = ASSET_ROOT_PATH;
    private static final String FONT_PATH = "textures/fonts/default";
    private static final int PIXELS_PER_EM = 24;

    private final PerspectiveCamera camera = new PerspectiveCamera();

    private final BlockDisplay blockDisplay;
    private final ModelDisplay modelDisplay;
    private final ParticleDisplay particleDisplay;
    private final SkyboxDisplay skyboxDisplay = new SkyboxDisplay();
    private final CloudDisplay cloudDisplay = new CloudDisplay();
    private final TextRenderer textRenderer;

    private final Queue<DrawTextCommand> drawTextCommandQueue;

    private int viewportWidth;
    private int viewportHeight;

    private int visibilityFlags = VisibilityFlags.DEFAULT;

    private Matrix4f cameraViewMatrix = new Matrix4f();
    private Matrix4f cameraProjectionMatrix = new Matrix4f();
    private Matrix4f textProjectionMatrix = new Matrix4f();

    private List<InvalidateColumnDto> columnsToInvalidate = new ArrayList<InvalidateColumnDto>();

    public CompositeDisplay(
            ClientParticleSpace clientParticleSpace,
            int viewportWidth,
            int viewportHeight)
    {
        blockDisplay = new BlockDisplay(camera);
        modelDisplay = new ModelDisplay(camera);
        particleDisplay = new ParticleDisplay(clientParticleSpace, camera);
        textRenderer = new TextRenderer(FONT_PATH);

        drawTextCommandQueue = new LinkedList<>();

        this.viewportWidth = viewportWidth;
        this.viewportHeight = viewportHeight;

        var mipmappedTextureAtlas = new TextureAtlas(TEXTURE_PATH, true);
        var textureAtlas = new TextureAtlas(TEXTURE_PATH, false);
        Assets.setup(mipmappedTextureAtlas, textureAtlas);
    }

    @Override
    public void close() {
        blockDisplay.close();
        Assets.tearDown();
    }

    public float getCameraX() { return camera.getX(); }
    public float getCameraY() { return camera.getY(); }
    public float getCameraZ() { return camera.getZ(); }
    public float getCameraYaw() { return camera.getYaw(); }
    public float getCameraPitch() { return camera.getPitch(); }
    public float getCameraFov() { return camera.getFov(); }
    public float getCameraWidth() { return camera.getWidth(); }
    public float getCameraHeight() { return camera.getHeight(); }
    public float getCameraNearClipPlane() { return camera.getNearClipPlane(); }
    public float getCameraFarClipPlane() { return camera.getFarClipPlane(); }
    public int getVisibilityFlags() { return visibilityFlags; }
    public void setVisibilityFlags(int to) { visibilityFlags = to; }
    public float getSkyboxBottomR() { return skyboxDisplay.getBottomR(); }
    public float getSkyboxBottomG() { return skyboxDisplay.getBottomG(); }
    public float getSkyboxBottomB() { return skyboxDisplay.getBottomB(); }
    public float getSkyboxTopR() { return skyboxDisplay.getTopR(); }
    public float getSkyboxTopG() { return skyboxDisplay.getTopG(); }
    public float getSkyboxTopB() { return skyboxDisplay.getTopB(); }
    public long getCloudSeed() { return cloudDisplay.getSeed(); }
    public float getCloudDensity() { return cloudDisplay.getDensity(); }
    public float getCloudElevation() { return cloudDisplay.getElevation(); }
    public float getCloudSpeedX() { return cloudDisplay.getSpeedX(); }
    public float getCloudSpeedZ() { return cloudDisplay.getSpeedZ(); }
    public float getCloudAlpha() { return cloudDisplay.getAlpha(); }

    public int getViewportWidth() { return viewportWidth; }
    public int getViewportHeight() { return viewportHeight; }
    public float getViewportAspectRatio() { return (float) viewportWidth / viewportHeight; }

    public boolean isVisible(int flag) { return (visibilityFlags & flag) != 0; }
    public void toggleVisibilityFlags(int flags) { visibilityFlags ^= flags; }

    public void tick(double elapsedTime) {
        cloudDisplay.tick(elapsedTime);
    }

    public void display() {
        cameraViewMatrix.identity();
        camera.getViewMatrix(cameraViewMatrix);
        cameraProjectionMatrix.identity();
        camera.getProjectionMatrix(cameraProjectionMatrix);

        Assets.getMipmappedTextureAtlas().updateTextures();
        Assets.getTextureAtlas().updateTextures();

        // Clears color and depth buffers.
        GLRenderContext.clearColorBuffer(0.0f, 0.0f, 0.0f, 0.0f);
        GLRenderContext.clearDepthBuffer(1.0f);

        // Display skybox.
        if (isVisible(VisibilityFlags.SKYBOX))
            skyboxDisplay.display(cameraViewMatrix, cameraProjectionMatrix);

        // Display blocks, entities, and clouds.
        GLState.pushFogEnabled(true);
        GLState.pushLinearFog(16.0f, 128.0f,
                skyboxDisplay.getMiddleR(),
                skyboxDisplay.getMiddleG(),
                skyboxDisplay.getMiddleB(),
                1.0f);
        {
            for (int i = 0; i < 6 && !columnsToInvalidate.isEmpty(); ++i) {
                var lastIndex = columnsToInvalidate.size() - 1;
                var message = columnsToInvalidate.remove(lastIndex);
                var columnPos = message.getColumnPos();
                var column = message.getColumn();
                blockDisplay.redrawColumn(columnPos, column);
            }

            for (int i = columnsToInvalidate.size() - 1; i >= 0; --i) {
                var message = columnsToInvalidate.get(i);
                if (message.getInvalidationPriority().equals(ColumnInvalidationPriority.NOW)) {
                    var columnPos = message.getColumnPos();
                    var column = message.getColumn();
                    blockDisplay.redrawColumn(columnPos, column);
                }
            }

            if (isVisible(VisibilityFlags.BLOCKS))
                blockDisplay.display();

            if (isVisible(VisibilityFlags.ENTITIES))
                modelDisplay.display();

            if (isVisible(VisibilityFlags.PARTICLES))
                particleDisplay.display();

            if (isVisible(VisibilityFlags.CLOUDS)) {
                cloudDisplay.display(
                        cameraViewMatrix,
                        cameraProjectionMatrix,
                        camera.getX(),
                        camera.getZ());
            }
        }
        GLState.popFogParams();
        GLState.popFogEnabled();

        // Draw Text
        if (isVisible(VisibilityFlags.HUD)) {
            textProjectionMatrix.identity();
            textProjectionMatrix.ortho2D(
                    0.0f,
                    (float) viewportWidth / PIXELS_PER_EM,
                    (float) viewportHeight / PIXELS_PER_EM,
                    0.0f);
            GLMatrixStack.setProjectionMatrix(textProjectionMatrix);

            GLMatrixStack.push();
            GLMatrixStack.loadIdentity();

            while (!drawTextCommandQueue.isEmpty()) {
                var command = drawTextCommandQueue.remove();
                var text = command.getText();
                var x = command.getX();
                var y = command.getY();
                var z = command.getZ();
                textRenderer.drawText(text, TextStyle.DEFAULT, x, y, z);
            }

            GLMatrixStack.pop();
        }
    }

    public void drawModel(
            double x, double y, double z,
            double yaw, double pitch, double roll,
            double scaleX, double scaleY, double scaleZ,
            ModelType modelType)
    {
        modelDisplay.drawModel(new DrawModelCommand(
                x, y, z,
                yaw, pitch, roll,
                scaleX, scaleY, scaleZ,
                modelType));
    }

    public void invalidateColumn(
            long columnPos,
            Column column,
            ColumnInvalidationPriority invalidationPriority)
    {
        columnsToInvalidate = columnsToInvalidate.stream()
                .filter(e -> e.getColumnPos() != columnPos)
                .collect(Collectors.toList());

        columnsToInvalidate.add(
                new InvalidateColumnDto(
                        columnPos,
                        column,
                        invalidationPriority));
    }

    public void moveCamera(float x, float y, float z) {
        camera.setX(x);
        camera.setY(y);
        camera.setZ(z);
    }

    public void orientCamera(float yaw, float pitch) {
        camera.setYaw(yaw);
        camera.setPitch(pitch);
        // TODO: Missing Camera::roll.
    }

    public void setCameraParams(float fov, float width, float height, float nearClipPlane, float farClipPlane) {
        camera.setFov(fov);
        camera.setWidth(width);
        camera.setHeight(height);
        camera.setNearClipPlane(nearClipPlane);
        camera.setFarClipPlane(farClipPlane);
    }

    public void setCloudParams(
            long seed,
            float density,
            float elevation,
            float speedX,
            float speedZ,
            float alpha)
    {
        cloudDisplay.setSeed(seed);
        cloudDisplay.setDensity(density);
        cloudDisplay.setElevation(elevation);
        cloudDisplay.setSpeed(speedX, speedZ);
        cloudDisplay.setAlpha(alpha);
    }

    public void setSkyColor(
            float br, float bg, float bb,
            float mr, float mg, float mb,
            float tr, float tg, float tb)
    {
        skyboxDisplay.setColor(br, bg, bb, mr, mg, mb, tr, tg, tb);
    }

    public void drawLabel(String text, double fontSize, double left, double top) {
        drawTextCommandQueue.add(new DrawTextCommand(text, left, top, 0.0));
    }
}

File: survival-graphics/src/main/java/net/survival/graphics/CompositeDisplay.kt
package net.survival.graphics

import net.survival.block.Column
import net.survival.graphics.opengl.GLMatrixStack
import net.survival.graphics.opengl.GLRenderContext
import net.survival.graphics.opengl.GLState
import net.survival.graphics.particle.ClientParticleSpace
import net.survival.render.ModelType
import org.joml.Matrix4f
import java.util.*
import java.util.stream.Collectors

private const val ASSET_ROOT_PATH = "./assets/"
private const val TEXTURE_PATH = ASSET_ROOT_PATH
private const val FONT_PATH = "textures/fonts/default"
private const val PIXELS_PER_EM = 24

class CompositeDisplay(
    clientParticleSpace: ClientParticleSpace?,
    viewportWidth: Int,
    viewportHeight: Int,
) {
    private val camera = PerspectiveCamera()

    private val blockDisplay: BlockDisplay
    private val modelDisplay: ModelDisplay
    private val particleDisplay: ParticleDisplay
    private val skyboxDisplay = SkyboxDisplay()
    private val cloudDisplay = CloudDisplay()
    private val textRenderer: TextRenderer

    private val drawTextCommandQueue: Queue<DrawTextCommand>

    val viewportWidth: Int
    val viewportHeight: Int

    var visibilityFlags = VisibilityFlags.DEFAULT

    private val cameraViewMatrix = Matrix4f()
    private val cameraProjectionMatrix = Matrix4f()
    private val textProjectionMatrix = Matrix4f()

    private var columnsToInvalidate: MutableList<InvalidateColumn> = ArrayList()

    init {
        blockDisplay = BlockDisplay(camera)
        modelDisplay = ModelDisplay(camera)
        particleDisplay = ParticleDisplay(clientParticleSpace!!, camera)
        textRenderer = TextRenderer(FONT_PATH)

        drawTextCommandQueue = LinkedList()

        this.viewportWidth = viewportWidth
        this.viewportHeight = viewportHeight

        val mipmappedTextureAtlas = TextureAtlas(TEXTURE_PATH, true)
        val textureAtlas = TextureAtlas(TEXTURE_PATH, false)
        Assets.setup(mipmappedTextureAtlas, textureAtlas)
    }

    fun close() {
        blockDisplay.close()
        Assets.tearDown()
    }

    val cameraX: Float get() = camera.x
    val cameraY: Float get() = camera.y
    val cameraZ: Float get() = camera.z
    val cameraYaw: Float get() = camera.yaw
    val cameraPitch: Float get() = camera.pitch
    val cameraFov: Float get() = camera.fov
    val cameraWidth: Float get() = camera.width
    val cameraHeight: Float get() = camera.height
    val cameraNearClipPlane: Float get() = camera.nearClipPlane
    val cameraFarClipPlane: Float get() = camera.farClipPlane
    val skyboxBottomR: Float get() = skyboxDisplay.bottomR
    val skyboxBottomG: Float get() = skyboxDisplay.bottomG
    val skyboxBottomB: Float get() = skyboxDisplay.bottomB
    val skyboxTopR: Float get() = skyboxDisplay.topR
    val skyboxTopG: Float get() = skyboxDisplay.topG
    val skyboxTopB: Float get() = skyboxDisplay.topB
    val cloudSeed: Long get() = cloudDisplay.seed
    val cloudDensity: Float get() = cloudDisplay.getDensity()
    val cloudElevation: Float get() = cloudDisplay.elevation
    val cloudSpeedX: Float get() = cloudDisplay.speedX
    val cloudSpeedZ: Float get() = cloudDisplay.speedZ
    val cloudAlpha: Float get() = cloudDisplay.alpha
    val viewportAspectRatio: Float get() = viewportWidth.toFloat() / viewportHeight

    fun isVisible(flag: Int): Boolean = visibilityFlags and flag != 0

    fun toggleVisibilityFlags(flags: Int) {
        visibilityFlags = visibilityFlags xor flags
    }

    fun tick(elapsedTime: Double) {
        cloudDisplay.tick(elapsedTime)
    }

    fun display() {
        cameraViewMatrix.identity()
        camera.getViewMatrix(cameraViewMatrix)
        cameraProjectionMatrix.identity()
        camera.getProjectionMatrix(cameraProjectionMatrix)

        Assets.getMipmappedTextureAtlas().updateTextures()
        Assets.getTextureAtlas().updateTextures()

        // Clears color and depth buffers.
        GLRenderContext.clearColorBuffer(0.0f, 0.0f, 0.0f, 0.0f)
        GLRenderContext.clearDepthBuffer(1.0f)

        // Display skybox.
        if (isVisible(VisibilityFlags.SKYBOX)) {
            skyboxDisplay.display(cameraViewMatrix, cameraProjectionMatrix)
        }

        // Display blocks, entities, and clouds.
        GLState.pushFogEnabled(true)
        GLState.pushLinearFog(16.0f, 128.0f,
            skyboxDisplay.middleR,
            skyboxDisplay.middleG,
            skyboxDisplay.middleB,
            1.0f
        )

        run {
            run {
                var i = 0
                while (i < 6 && columnsToInvalidate.isNotEmpty()) {
                    val lastIndex = columnsToInvalidate.size - 1
                    val message = columnsToInvalidate.removeAt(lastIndex)
                    val columnPos = message.columnPos
                    val column = message.column
                    blockDisplay.redrawColumn(columnPos, column)
                    ++i
                }
            }

            for (i in columnsToInvalidate.indices.reversed()) {
                val message = columnsToInvalidate[i]
                if (message.invalidationPriority == ColumnInvalidationPriority.NOW) {
                    val columnPos = message.columnPos
                    val column = message.column
                    blockDisplay.redrawColumn(columnPos, column)
                }
            }

            if (isVisible(VisibilityFlags.BLOCKS)) {
                blockDisplay.display()
            }

            if (isVisible(VisibilityFlags.ENTITIES)) {
                modelDisplay.display()
            }

            if (isVisible(VisibilityFlags.PARTICLES)) {
                particleDisplay.display()
            }

            if (isVisible(VisibilityFlags.CLOUDS)) {
                cloudDisplay.display(
                    cameraViewMatrix,
                    cameraProjectionMatrix,
                    camera.x,
                    camera.z
                )
            }
        }

        GLState.popFogParams()
        GLState.popFogEnabled()

        // Draw Text
        if (isVisible(VisibilityFlags.HUD)) {
            textProjectionMatrix.identity()
            textProjectionMatrix.ortho2D(
                0.0f,
                viewportWidth.toFloat() / PIXELS_PER_EM,
                viewportHeight.toFloat() / PIXELS_PER_EM,
                0.0f
            )

            GLMatrixStack.setProjectionMatrix(textProjectionMatrix)
            GLMatrixStack.push()
            GLMatrixStack.loadIdentity()

            while (!drawTextCommandQueue.isEmpty()) {
                val command = drawTextCommandQueue.remove()
                val text = command.text
                val x = command.x
                val y = command.y
                val z = command.z
                textRenderer.drawText(text, TextStyle.DEFAULT, x, y, z)
            }

            GLMatrixStack.pop()
        }
    }

    fun drawModel(
        x: Double, y: Double, z: Double,
        yaw: Double, pitch: Double, roll: Double,
        scaleX: Double, scaleY: Double, scaleZ: Double,
        modelType: ModelType,
    ) {
        modelDisplay.drawModel(DrawModelCommand(
            x, y, z,
            yaw, pitch, roll,
            scaleX, scaleY, scaleZ,
            modelType
        ))
    }

    fun invalidateColumn(
        columnPos: Long,
        column: Column?,
        invalidationPriority: ColumnInvalidationPriority,
    ) {
        columnsToInvalidate = columnsToInvalidate.stream()
            .filter { e: InvalidateColumn -> e.columnPos != columnPos }
            .collect(Collectors.toList())
        columnsToInvalidate.add(InvalidateColumn(columnPos, column, invalidationPriority))
    }

    fun moveCamera(x: Float, y: Float, z: Float) {
        camera.x = x
        camera.y = y
        camera.z = z
    }

    fun orientCamera(yaw: Float, pitch: Float) {
        camera.yaw = yaw
        camera.pitch = pitch
        // TODO: Missing Camera::roll.
    }

    fun setCameraParams(fov: Float, width: Float, height: Float, nearClipPlane: Float, farClipPlane: Float) {
        camera.fov = fov
        camera.width = width
        camera.height = height
        camera.nearClipPlane = nearClipPlane
        camera.farClipPlane = farClipPlane
    }

    fun setCloudParams(
        seed: Long,
        density: Float,
        elevation: Float,
        speedX: Float,
        speedZ: Float,
        alpha: Float,
    ) {
        cloudDisplay.seed = seed
        cloudDisplay.setDensity(density)
        cloudDisplay.elevation = elevation
        cloudDisplay.speedX = speedX
        cloudDisplay.speedZ = speedZ
        cloudDisplay.alpha = alpha
    }

    fun setSkyColor(
        br: Float, bg: Float, bb: Float,
        mr: Float, mg: Float, mb: Float,
        tr: Float, tg: Float, tb: Float,
    ) {
        skyboxDisplay.setColor(br, bg, bb, mr, mg, mb, tr, tg, tb)
    }

    fun drawLabel(text: String, fontSize: Double, left: Double, top: Double) {
        drawTextCommandQueue.add(DrawTextCommand(text, left, top, 0.0))
    }
}

object VisibilityFlags {
    const val BLOCKS = 0x1
    const val ENTITIES = 0x2
    const val PARTICLES = 0x4
    const val SKYBOX = 0x8
    const val CLOUDS = 0x10
    const val HUD = 0x20
    const val DEBUG_GEOMETRY = 0x40
    const val DEFAULT = BLOCKS or ENTITIES or PARTICLES or SKYBOX or CLOUDS or HUD or DEBUG_GEOMETRY
}

enum class ColumnInvalidationPriority {
    LOW, NOW
}

private class InvalidateColumn(
    val columnPos: Long,
    val column: Column?,
    val invalidationPriority: ColumnInvalidationPriority
)

internal sealed class DrawCommand

internal class DrawTextCommand(val text: String, val x: Double, val y: Double, val z: Double): DrawCommand()

internal class DrawModelCommand(
    val x: Double, val y: Double, val z: Double,
    val yaw: Double, val pitch: Double, val roll: Double,
    val scaleX: Double, val scaleY: Double, val scaleZ: Double,
    val modelType: ModelType,
): DrawCommand()
